README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# whoville <a href='https://github.com/gpw13/whoville'><img src='man/figures/logo.png' align="right" height="139" /></a>

<!-- badges: start -->

[![R build
status](https://github.com/gpw13/whoville/workflows/R-CMD-check/badge.svg)](https://github.com/gpw13/whoville/actions)
<!-- badges: end -->

The goal of whoville is to provide a package country reference data
published by the World Health Organization, United Nations and World
Bank easily accessible in R. At the moment, this is a reference dataset
and set of functions to work with country codes and names, allowing easy
conversion between names and codes as well as easy access to region
codes, WHO member status, and other published country metadata. This is
used to assist the work of the WHO’s Division of Data, Analytics, and
Delivery for Impact’s Global Programme of Work 13.

## Functions

Some functions are built to help you turn names or country codes into
ISO3 codes:

-   `names_to_code()` is the workhorse function that helps match country
    names and returns standardized country codes.
-   `names_to_iso3()` is a wrapper around `names_to_code()` that
    automatically outputs ISO3 code.
-   `codes_to_iso3()` converts other country codes into ISO3 format.
-   `valid_codes()` lets you quickly check which country codes in a
    vector are recognized by the whoville package.

Other functions encourage a tidy R workflow where ISO3 codes are used as
the unique identifier for each country:

-   `iso3_to_regions()` takes in a vector of ISO3 codes and returns
    specified region values.
-   `iso3_to_codes()` takes in a vector of ISO3 codes and returns
    specified country codes.
-   `iso3_to_names()` takes in a vector of ISO3 codes and returns
    specified country names.
-   `is_who_member()` takes in a vector of ISO3 codes and checks if they
    are a WHO member state or not.
-   `is_oecd_member()` takes in a vector of ISO3 codes and checks if
    they are an OECD member state or not.
-   `is_gbd_high_income()` takes in a vector of ISO3 codes and checks if
    they are classified as high-income in the 2019 GBD from IHME.

All of these functions are built on top of the `countries` data frame
also exported with the package and developed off of public datasets
provided by the World Health Organization and United Nations. Details
available through `?countries`.

## Installation

You can install whoville from Github with:

``` r
devtools::install_github("gpw13/whoville")
```

## Usage

If we have an unclean data frame with country names, we can use
`names_to_codes()` to match these to ISO3 codes. The function matches
the names vector across all possible names found in the `countries` data
frame. ISO3 codes for exact matches are always returned, but the user
has specific options for non-exact matches. They can be fuzzy matched
(the default), always made NA, or require user input to confirm fuzzy
matching results. Fuzzy matches always return a message to the user on
the confirmed match. More details available through `?names_to_codes`.

``` r
library(whoville)

names_to_code(c("Venezuela", "Arentina", "afghanist"))
#> 'arentina' has no exact match. Closest name found was 'argentina'.
#> 'afghanist' has no exact match. Closest name found was 'afghanistan'.
#> [1] "VEN" "ARG" "AFG"
```

Since these functions are vectorized, we can easily use them in a normal
workflow, especially if we’re using the `tidyverse`. Below, we can clean
up our tidy names and get the correct UN region and income group for our
countries, as well as its name in Chinese:

``` r
library(dplyr)
df <- data.frame(c_names = c("Venezuela", "Arentina", "afghanist"))

df %>%
  mutate(iso3 = names_to_code(c_names),
         un_region = iso3_to_regions(iso3, region = "un_region"),
         wb_ig = iso3_to_regions(iso3, region = "wb_ig"),
         name_zh = iso3_to_names(iso3, org = "un", language = "zh"))
#>     c_names iso3 un_region wb_ig                name_zh
#> 1 Venezuela  VEN        19   UMC 委内瑞拉玻利瓦尔共和国
#> 2  Arentina  ARG        19   UMC                 阿根廷
#> 3 afghanist  AFG       142   LIC                 阿富汗
```


NEWS.md
# whoville 0.1.7

* IHME's GBD codes and regional classifications added to allow conversion between ISO3 and GBD codes, as well as use of
the GBD regions in analyses.

# whoville 0.1.6

* Minor change to accommodate new formatting for the World Bank's historical classification by income for countries.

# whoville 0.1.5

* Added UN intermediate regions, UN DESA regions and subregions, and SDG regions
and subregions.

# whoville 0.1.4

* Small island developing states (SIDS) and land-locked developing countries (LLDC)
    classifications added to `countries`.
* Additional alt names from official ISO3 short names added.

# whoville 0.1.3

* UN region and subregion names added in for every language, which can be accessed
    using `iso3_to_regions()`.
* OECD membership added, usable through `is_oecd_member()` and `oecd_member_states().
* GBD 2019 high-income countries added, usable through `is_gbd_high_income()` and
    `gbd_high_income_states()`.
* UN least-developed country status added, usable through `is_un_ldc()` and 
    `un_ldcs()`.

# whotilities 0.1.2

* Country data updated to use WHO xMart4 REF_MART data. This has created changes across all functions and in the package.
* WHO formal and short names now available in all languages
* Reduced the number of country codes available to ISO3, ISO2, ISO (numeric), WHO code, and M49
* Increased numbers of countries to 248
* Accessed World Bank Income Group data directly from World Bank website, now have income group classifications for 1987 - 2019 (corresponds to fiscal year 1989 - 2021)

# whotilities 0.1.1

* Added a `NEWS.md` file to track changes to the package.
* Added `who_member_states()` to quickly extract ISO3 codes for all WHO member states

# whotilities 0.1.0

* Initial release of package
* Publication of `countries` dataframe with codes, names, and regional categorizations
* Added functions to coerce names to codes (and vice-versa) released
* Added functions to get key country information using ISO3 codes


R/data.R
#' Country metadata from the WHO, UN, and WB.
#'
#' A dataset containing names, codes, and other attributes
#' for countries included in World Health Organization,
#' United Nations and World Bank datasets.
#'
#' @format A data frame with `r nrow(countries)` rows and `r ncol(countries)` variables:
#' \describe{
#'   \item{iso3}{\href{https://en.wikipedia.org/wiki/ISO_3166-1_alpha-3}{ISO3 code.}}
#'   \item{iso2}{\href{https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2}{ISO2 code.}}
#'   \item{iso_numeric}{\href{https://en.wikipedia.org/wiki/ISO_3166-1_numeric}{ISO numeric code.}}
#'   \item{who_code}{\href{https://apps.who.int/gho/data/node.metadata.COUNTRY?lang=en}{World Health Organization code.}}
#'   \item{m49}{\href{https://en.wikipedia.org/wiki/UN_M49}{United Nations M49 code.}}
#'   \item{gbd_code}{\href{http://www.healthdata.org/gbd/faq}{IHME GBD country code.}}
#'   \item{sovereign_iso3}{ISO3 code of sovereign nation. Value is `NA` if sovereign itself.}
#'   \item{who_member}{Logical, is country a WHO member state?}
#'   \item{who_member_small}{Logical, is country a WHO member state with population <= 90,000?}
#'   \item{oecd_member}{Logical, is country an OECD member state?}
#'   \item{gbd_high_income}{Logical, is the country classified as high-income in the IHME's GBD 2019?}
#'   \item{un_ldc}{Logical, is the country classified as a least-developed country by the United Nations?}
#'   \item{un_lldc}{Logical, is the country classified as a land-locked developing country by the United Nations?}
#'   \item{un_sids}{Logical, is the country classified as a small island developing state by the United Nations?}
#'   \item{who_short_name_en}{WHO country name, short, English.}
#'   \item{who_formal_name_en}{WHO country name, formal, English. Only available for official member states.}
#'   \item{who_short_name_ar}{WHO country name, short, Arabic.}
#'   \item{who_formal_name_ar}{WHO country name, formal, Arabic. Only available for official member states.}
#'   \item{who_short_name_es}{WHO country name, short, Spanish.}
#'   \item{who_formal_name_es}{WHO country name, formal, Spanish. Only available for official member states.}
#'   \item{who_short_name_fr}{WHO country name, short, French.}
#'   \item{who_formal_name_fr}{WHO country name, formal, French. Only available for official member states.}
#'   \item{who_short_name_ru}{WHO country name, short, Russian.}
#'   \item{who_formal_name_ru}{WHO country name, formal, Russian. Only available for official member states.}
#'   \item{who_short_name_zh}{WHO country name, short, Chinese.}
#'   \item{who_formal_name_zh}{WHO country name, formal, Chinese. Only available for official member states.}
#'   \item{un_name_en}{UN country name, English.}
#'   \item{un_name_ru}{UN country name, Russian.}
#'   \item{un_name_fr}{UN country name, French.}
#'   \item{un_name_es}{UN country name, Spanish.}
#'   \item{un_name_ar}{UN country name, Arabic.}
#'   \item{un_name_zh}{UN country name, Chinese.}
#'   \item{alt_name_en}{Alternative (non-official) country name, English.}
#'   \item{alt_name_2_en}{Another alternative (non-official) country name, English.}
#'   \item{alt_name_3_en}{Another alternative (non-official) country name, English.}
#'   \item{alt_name_4_en}{Another alternative (non-official) country name, English.}
#'   \item{alt_name_5_en}{Another alternative (non-official) country name, English.}
#'   \item{former_name_en}{Former country (non-official) name, English.}
#'   \item{former_name_2_en}{Another former (non-official) country name, English.}
#'   \item{who_region}{WHO region code.}
#'   \item{un_region}{UN region code.}
#'   \item{un_subregion}{UN sub-region code.}
#'   \item{un_intermediate_region}{UN intermediate region code.}
#'   \item{un_region_name_en}{UN region name, English.}
#'   \item{un_subregion_name_en}{UN subregion name, English.}
#'   \item{un_intermediate_region_name_en}{UN intermediate region name, English.}
#'   \item{un_region_name_ru}{UN region name, Russian.}
#'   \item{un_subregion_name_ru}{UN subregion name, Russian.}
#'   \item{un_intermediate_region_name_ru}{UN intermediate region name, Russian.}
#'   \item{un_region_name_fr}{UN region name, French.}
#'   \item{un_subregion_name_fr}{UN subregion name, French.}
#'   \item{un_intermediate_region_name_fr}{UN intermediate region name, French.}
#'   \item{un_region_name_es}{UN region name, Spanish.}
#'   \item{un_subregion_name_es}{UN subregion name, Spanish.}
#'   \item{un_intermediate_region_name_es}{UN intermediate region name, Spanish.}
#'   \item{un_region_name_ar}{UN region name, Arabic.}
#'   \item{un_subregion_name_ar}{UN subregion name, Arabic.}
#'   \item{un_intermediate_region_name_ar}{UN intermediate region name, Arabic.}
#'   \item{un_region_name_zh}{UN region name, Chinese.}
#'   \item{un_subregion_name_zh}{UN subregion name, Chinese.}
#'   \item{un_intermediate_region_name_zh}{UN intermediate region name, Chinese.}
#'   \item{un_desa_region}{UN DESA region code.}
#'   \item{un_desa_region_name_en}{UN DESA region name, English.}
#'   \item{un_desa_subregion}{UN DESA subregion code.}
#'   \item{un_desa_subregion_name_en}{UN DESA subregion name, English.}
#'   \item{sdg_region}{SDG region code.}
#'   \item{sdg_region_name_en}{SDG region name, English.}
#'   \item{sdg_subregion}{SDG subregion code.}
#'   \item{sdg_subregion_name_en}{SDG subregion name, English.}
#'   \item{gbd_region}{IHME GBD region code.}
#'   \item{gbd_region_name_en}{IHME GBD region name, English.}
#'   \item{gbd_subregion}{IHME GBD subregion code.}
#'   \item{gbd_subregion_name_en}{IHME GBD subregion name, English.}
#'   \item{wb_region}{WB geographical region. English name as no code provided by WB.}
#'   \item{wb_region_name_en}{WB geographical region name, English.}
#'   \item{wb_ig_1987}{World Bank Income Group code, 1987 (FY 1989) classification.}
#'   \item{wb_ig_1988}{World Bank Income Group code, 1988 (FY 1990) classification.}
#'   \item{wb_ig_1989}{World Bank Income Group code, 1989 (FY 1991) classification.}
#'   \item{wb_ig_1990}{World Bank Income Group code, 1990 (FY 1992) classification.}
#'   \item{wb_ig_1991}{World Bank Income Group code, 1991 (FY 1993) classification.}
#'   \item{wb_ig_1992}{World Bank Income Group code, 1992 (FY 1994) classification.}
#'   \item{wb_ig_1993}{World Bank Income Group code, 1993 (FY 1995) classification.}
#'   \item{wb_ig_1994}{World Bank Income Group code, 1994 (FY 1996) classification.}
#'   \item{wb_ig_1995}{World Bank Income Group code, 1995 (FY 1997) classification.}
#'   \item{wb_ig_1996}{World Bank Income Group code, 1996 (FY 1998) classification.}
#'   \item{wb_ig_1997}{World Bank Income Group code, 1997 (FY 1999) classification.}
#'   \item{wb_ig_1998}{World Bank Income Group code, 1998 (FY 2000) classification.}
#'   \item{wb_ig_1999}{World Bank Income Group code, 1999 (FY 2001) classification.}
#'   \item{wb_ig_2000}{World Bank Income Group code, 2000 (FY 2002) classification.}
#'   \item{wb_ig_2001}{World Bank Income Group code, 2001 (FY 2003) classification.}
#'   \item{wb_ig_2002}{World Bank Income Group code, 2002 (FY 2004) classification.}
#'   \item{wb_ig_2003}{World Bank Income Group code, 2003 (FY 2005) classification.}
#'   \item{wb_ig_2004}{World Bank Income Group code, 2004 (FY 2006) classification.}
#'   \item{wb_ig_2005}{World Bank Income Group code, 2005 (FY 2007) classification.}
#'   \item{wb_ig_2006}{World Bank Income Group code, 2006 (FY 2008) classification.}
#'   \item{wb_ig_2007}{World Bank Income Group code, 2007 (FY 2009) classification.}
#'   \item{wb_ig_2008}{World Bank Income Group code, 2008 (FY 2010) classification.}
#'   \item{wb_ig_2009}{World Bank Income Group code, 2009 (FY 2011) classification.}
#'   \item{wb_ig_2010}{World Bank Income Group code, 2010 (FY 2012) classification.}
#'   \item{wb_ig_2011}{World Bank Income Group code, 2011 (FY 2013) classification.}
#'   \item{wb_ig_2012}{World Bank Income Group code, 2012 (FY 2014) classification.}
#'   \item{wb_ig_2013}{World Bank Income Group code, 2013 (FY 2015) classification.}
#'   \item{wb_ig_2014}{World Bank Income Group code, 2014 (FY 2016) classification.}
#'   \item{wb_ig_2015}{World Bank Income Group code, 2015 (FY 2017) classification.}
#'   \item{wb_ig_2016}{World Bank Income Group code, 2016 (FY 2018) classification.}
#'   \item{wb_ig_2017}{World Bank Income Group code, 2017 (FY 2019) classification.}
#'   \item{wb_ig_2018}{World Bank Income Group code, 2018 (FY 2020) classification.}
#'   \item{wb_ig_2019}{World Bank Income Group code, 2019 (FY 2021) classification.}
#'   \item{wb_ig_2020}{World Bank Income Group code, 2020 (FY 2022) classification.}
#' }
#' @source \href{https://portal-uat.who.int/xmart4/REFMART}{World Health Organization country reference table, accessed September 23 2020.}
#' @source \href{https://unstats.un.org/unsd/methodology/m49/overview/}{United Nations Statistics Division, Country or Area Codes for Statistical Use tables; accessed August 15 2020.}
#' @source \href{https://datahelpdesk.worldbank.org/knowledgebase/articles/906519-world-bank-country-and-lending-groups}{World Bank income groups, fiscal years 1989 - 2021, and regions, accessed August 15 2020.}
#' @source \href{https://www.oecd.org/about/members-and-partners/}{Organisation for Economic Co-operation and Development membership list, accessed 1 February 2021.}
#' @source \href{http://ghdx.healthdata.org/gbd-2019}{Institute for Health Metrics and Evaluation Global Burden of Disease 2019 Socio-Demographic Index data, accessed 1 February 2021.}
#' @source \href{https://www.un.org/development/desa/dpad/wp-content/uploads/sites/45/publication/ldc_list.pdf}{United Nations Committee for Development Policy, list of least developed countries, accessed 16 February 2021.}
#' @source \href{https://www.un.org/development/desa/pd/sites/www.un.org.development.desa.pd/files/aggregates_correspondence_table_2020_1.xlsx}{United Nations Department of Economic and Social Affairs, Population Division (2020). International Migrant Stock 2020.}
#' @source \href{http://ghdx.healthdata.org/sites/default/files/ihme_query_tool/IHME_GBD_2019_CODEBOOK.zip}{Global Burden of Disease 2019 country and regional classifications, Institute for Health Metrics and Evaluation.}
"countries"



R/codes.R
#' Get regions from ISO3 country codes.
#'
#' `iso3_to_regions()` takes in a vector of ISO3 codes and returns a vector of
#' specified regions. Currently, this can be codes or English names for
#' UN regions/subregions/intermediate regions, UN DESA regions/subregions,
#' SDG regions/subregions, WHO regions, and World Bank geographical regions and
#' Income Groups.
#'
#' @param iso3 Character vector of ISO3 codes.
#' @param region Type of region to return. Regional classifications available from
#'     the WHO, UN, UNSD SDGs, UN DESA, the World Bank (geographic and income based),
#'     and IHME's GBD. See the [whoville::countries] documentation for more source
#'     details.
#' @param year For World Bank Income Group only, specify the year to return.
#' @param name For UN, UNSD SDG, UN DESA, WB and IHME GBD regions only. If `TRUE`,
#'     return official regional name instead of code.
#' @param language For `"un_region"`, `"un_subregion"`, and `"un_intermediate_region"`
#'     only if returning name. A character value specifying the language of the
#'     country names. Should be specified using the ISO2 language code.
#'
#' @return Character vector.
#'
#' @export
iso3_to_regions <- function(iso3,
                            region = c("who_region", "un_region", "un_subregion", "un_intermediate_region", "sdg_region", "sdg_subregion", "gbd_region", "gbd_subregion", "un_desa_region", "un_desa_subregion", "wb_region", "wb_ig"),
                            year = max(wb_ig_years()),
                            name = FALSE,
                            language = c("en", "es", "ru", "ar", "zh", "fr")) {
  region <- rlang::arg_match(region)

  if (region == "wb_ig") {
    assert_wb_ig_years(year)
    mtch <- paste(region, year, sep = "_")
  } else if (region %in% c("un_region", "un_subregion", "un_intermediate_region", "sdg_region", "sdg_subregion", "gbd_region", "gbd_subregion", "un_desa_region", "un_desa_subregion", "wb_region") && name) {
    language <- rlang::arg_match(language)
    mtch <- paste(region, "name", language, sep = "_")
  } else {
    mtch <- region
  }

  regions <- whoville::countries[[mtch]]
  idx <- match(iso3, whoville::countries[["iso3"]])
  regions[idx]
}

#' Check if country codes are present in `countries` data frame.
#'
#' `valid_codes()` takes in a vector of country codes and returns a logical vector
#' on whether that country code is present in `countries[["iso3"]]`
#'
#' @param codes Character vector of country codes.
#' @param type A character value specifying the type of country code supplied.
#' All possible values available through `country_code_types()`.
#'
#' @return Logical vector.
#'
#' @export
valid_codes <- function(codes, type = "iso3") {
  rlang::arg_match(type, country_code_types())
  codes %in% whoville::countries[["type"]]
}

#' Get country names from ISO3 country codes.
#'
#' `iso3_to_names()` takes in a vector of ISO3 codes and returns names as
#' specified by the user. Currently, this can be World Health Organization names,
#' both short and formal, in all 6 official languages and United Nations names in
#' all 6 official languages.
#'
#' @param iso3 Character vector of ISO3 codes.
#' @param org Official names of which organization to return. Can be
#' either "who" (default) or "un".
#' @param type "short" (default) or "formal" name to return. Only used when org is "who".
#' @param language A character value specifying the language of the country names.
#' Should be specified using the ISO2 language code. Defaults to "en", but matching
#' available for all 6 official WHO languages. Possible values are "en", "es",
#' "ar", "fr", "ru", and "zh".
#'
#' @return Character vector.
#'
#' @export
iso3_to_names <- function(iso3,
                          org = c("who", "un"),
                          type = c("short", "formal"),
                          language = c("en", "es", "ru", "ar", "zh", "fr")) {
  org <- rlang::arg_match(org)
  language <- rlang::arg_match(language)
  if (org == "who") {
    type <- rlang::arg_match(type)
    org <- paste(org, type, sep = "_")
  }

  rgx <- sprintf("^%s_name_%s$", org, language)
  names <- whoville::countries[[which(grepl(rgx, names(whoville::countries)))]]
  idx <- match(iso3, whoville::countries[["iso3"]])
  names[idx]
}

#' Get country codes from ISO3 country codes.
#'
#' `iso3_to_codes()` takes in a vector of ISO3 codes and returns a country code
#' vector specified by the user.
#'
#' @param iso3 Character vector of ISO3 codes.
#' @param type A character value specifying the type of country code to return.
#' All possible values available through `country_code_types()`.
#'
#' @return Character vector.
#'
#' @export
iso3_to_codes <- function(iso3, type) {
  rlang::arg_match(type, country_code_types())

  codes <- whoville::countries[[type]]
  idx <- match(iso3, whoville::countries[["iso3"]])
  codes[idx]
}

#' Get ISO3 codes from other country codes.
#'
#' `codes_to_iso3()` takes in a vector of country codes and returns ISO3 codes.
#'
#' @param codes Character vector of country codes.
#' @inheritParams iso3_to_codes
#'
#' @return Character vector.
#'
#' @export
codes_to_iso3 <- function(codes, type) {
  rlang::arg_match(type, country_code_types())

  iso3 <- whoville::countries[["iso3"]]
  idx <- match(codes, whoville::countries[[type]])
  iso3[idx]
}


R/names.R
#' Get country codes from names.
#'
#' `names_to_codes()` takes in a vector of country names and returns a vector of
#' specified country codes.
#'
#' The function first searches for exact matches in the `countries`
#' data frame, and if none are found, uses fuzzy matching to guess the country.
#' Matching is done using `stringdist` [fuzzy matching methods][stringdist::stringdist-metrics],
#' although `names_to_codes()` allows the user to not do any fuzzy matching (non-exact
#' matches returned as `NA`) or require user input.
#'
#' @param names A character vector of country names.
#' @param type A character value specifying the type of country code to return.
#'     Defaults to ISO3. All possible values available through `country_code_types()`.
#' @param language A character value specifying the language of the country names.
#'     Should be specified using the ISO2 language code. Defaults to "en", but matching
#'     available for all 6 official WHO languages. Possible values are "en", "es",
#'     "ar", "fr", "ru", and "zh".
#' @param ignore_case A logical value specifying whether or not to ignore case when
#'     string matching. Defaults to TRUE.
#' @param fuzzy_matching A character value specifying whether or not to do fuzzy matching,
#'     or to require user input when no exact match is found. Defaults to "yes".
#' @param method A character value specifying the distance method for use in fuzzy matching
#'     if an exact match is not found. Defaults to Jaro-Winker due to its value
#'     in matching names. See [stringdist::stringdist()] for all possible methods.
#' @param p A numeric value between 0 and 0.25 specifying the penalty factor
#'     for Jaro-Winkler distance. Ignored if method is not "jw".
#'
#' @return A vector of country codes matched to the names.
#'
#' @export
names_to_code <- function(names,
                          type = "iso3",
                          language = c("en", "es", "ru", "ar", "zh", "fr"),
                          ignore_case = T,
                          fuzzy_matching = c("yes", "no", "user_input"),
                          method = "jw",
                          p = 0.1) {
  rlang::arg_match(type, country_code_types())
  language <- rlang::arg_match(language)
  assert_logical(ignore_case)
  fuzzy_matching <- assert_fuzzy_matching(fuzzy_matching)
  rlang::arg_match(method, c("osa", "lv", "dl", "hamming", "lcs", "qgram", "cosine", "jaccard", "jw", "soundex"))
  assert_p(p)

  df <- dplyr::select(whoville::countries[,name_cols()],
                      dplyr::ends_with(paste0("_", language)))

  if (ignore_case) {
    df <- dplyr::mutate(df,
                        dplyr::across(dplyr::everything(),
                                      tolower))
    names <- tolower(names)
  }

  # drop missing values in names
  names <- names[!is.na(names)]

  result <- sapply(unique(names),
                   name_matching,
                   df = df,
                   method = method,
                   p = p,
                   fm = fuzzy_matching,
                   type = type)

  unname(result[names])
}

#' Get ISO3  codes from names
#'
#' `names_to_iso3()` takes in a vector of country names and returns a vector of
#' ISO3 codes.
#'
#' @inherit names_to_code details
#' @inherit names_to_code return
#' @inheritParams names_to_code
#'
#' @export
names_to_iso3 <- function(names,
                          language = c("en", "es", "ru", "ar", "zh", "fr"),
                          ignore_case = T,
                          fuzzy_matching = c("yes", "no", "user_input"),
                          method = "jw",
                          p = 0.1) {
  names_to_code(names = names,
                type = "iso3",
                language = language,
                ignore_case = ignore_case,
                fuzzy_matching = fuzzy_matching,
                method = method,
                p = p)
}

#' @noRd
name_matching <- function(name,
                          df,
                          method,
                          p,
                          fm,
                          type) {
  scrs <- apply(df, 2, stringdist::stringdist, name, method = method, p = p)
  scr_mins <- apply(scrs, 1, function(x) suppressWarnings(min(x, na.rm = T)))
  row <- which.min(scr_mins)
  col <- apply(scrs, 1, which.min)[[row]]
  fit <- df[row, col]
  fuzz_result <- whoville::countries[[type]][row]
  if (min(scr_mins) != 0) {
    if (fm == "no") {
      result <- NA_character_
    } else {
      if (fm == "yes") {
        message("'", name, "' has no exact match. Closest name found was '", fit, "'.")
        result <- fuzz_result
      } else {
        check <- T
        while (check) {
          message("The best match found for '", name, "' is '", fit, "' with ", toupper(type)," code ", fuzz_result, ".")
          result <- readline(sprintf("Confirm the correct %s code for %s. Type 'N/A' to skip this country: ",
                                     toupper(type),
                                     name))
          if (result %in% c("N/A", whoville::countries[[type]])) {
            check <- F
            if (result == "N/A") result <- NA_character_
          } else {
            check_input <- readline(sprintf("%s is not a valid %s code found in the countries file. Please confirm if %s is correct (y/n): ",
                                            result,
                                            toupper(type),
                                            result))
            if (tolower(check_input) == "y") {
              check <- F
            } else {
              message("Since not 'y', repeating entry for ", name, ".")
            }
          }
        }
      }
    }
  } else {
    result <- fuzz_result
  }
  result
}

#' @noRd
assert_p <- function(p) {
  lngth <- length(p)
  if (lngth > 1) {
    stop(sprintf("p must be of length 1, not %s",
                 lngth),
         call. = FALSE)
  }

  cls <- class(p)
  if (!is.numeric(p)) {
    stop(sprintf("p must be a numeric value, not %s",
                 cls),
         call. = FALSE)
  }

  if (!dplyr::between(p, 0, 0.25)) {
    stop(sprintf("p must be between 0 and 0.25, not %s",
                 p),
         call. = FALSE)
  }
}

assert_fuzzy_matching <- function(fuzzy_matching = c("yes", "no", "user_input")) {
  fuzzy_matching <- rlang::arg_match(fuzzy_matching)
  if (fuzzy_matching == "user_input" & !interactive()) {
    stop("R session must be interactive if fuzzying_matching is set to 'user_input'. See ?interactive for more details.")
  }

  fuzzy_matching
}


R/member_states.R
#' Get WHO membership status from ISO3 codes.
#'
#' `is_who_member()` takes in a vector of ISO3 codes and returns a logical vector
#' on whether that country is a WHO member state or not.
#'
#' @param iso3 Character vector of ISO3 codes.
#'
#' @return Logical vector.
#'
#' @export
is_who_member <- function(iso3) {
  members <- whoville::countries[["who_member"]]
  idx <- match(iso3, whoville::countries[["iso3"]])
  members[idx] %in% TRUE
}


#' Get WHO small state membership status from ISO3 codes.
#'
#' `is_who_member_small()` takes in a vector of ISO3 codes and returns a logical vector
#' on whether that country is a small WHO member state or not. In this instance,
#' small member states are defined as member states with a total population <=
#' 90,000 in mid-year 2018, according to the World Population Prospects.
#'
#' @inheritParams is_who_member
#'
#' @return Logical vector.
#'
#' @export
is_who_member_small <- function(iso3) {
  members <- whoville::countries[["who_member_small"]]
  idx <- match(iso3, whoville::countries[["iso3"]])
  members[idx] %in% TRUE
}

#' Get large state membership status from ISO3 codes.
#'
#' `is_who_member_large()` takes in a vector of ISO3 codes and returns a logical vector
#' on whether that country is a large WHO member state or not. In this instance,
#' large member states are defined as member states with a total population >
#' 90,000 in mid-year 2018, according to the World Population Prospects.
#'
#' @inheritParams is_who_member
#'
#' @return Logical vector.
#'
#' @export
is_who_member_large <- function(iso3) {
  is_who_member(iso3) & !is_who_member_small(iso3)
}

#' Get ISO3 codes for WHO member states.
#'
#' `who_member_states()` returns ISO3 codes for WHO member states. By default,
#' it returns all countries, but can be subset to only small (less than 90,000
#' population) or large states. Useful to expand data frames to explicitly include
#' missing data for countries.
#'
#' @param include Subset of member states to return.
#'     * "all": All member states (the default)
#'     * "small": Small member states (2018 population < 90,000)
#'     * "large": Large member states (2018 population >= 90,000)
#'
#' @return A character vector of ISO3 codes.
#'
#' @export
who_member_states <- function(include = c("all", "small", "large")) {
  include <- rlang::arg_match(include)
  x <- whoville::countries[["iso3"]]
  if (include == "small") {
    x <- x[is_who_member_small(x)]
  } else if (include == "large") {
    x <- x[is_who_member_large(x)]
  } else {
    x <- x[is_who_member(x)]
  }
  x
}

#' Get OECD membership status from ISO3 codes.
#'
#' `is_oecd_member()` takes in a vector of ISO3 codes and returns a logical vector
#' on whether that country is an OECD member state or not.
#'
#' @inheritParams is_who_member
#'
#' @return Logical vector.
#'
#' @export
is_oecd_member <- function(iso3) {
  members <- whoville::countries[["oecd_member"]]
  idx <- match(iso3, whoville::countries[["iso3"]])
  members[idx] %in% TRUE
}

#' Get ISO3 codes for OECD member states.
#'
#' `oecd_member_states()` returns ISO3 codes for OECD member states. Useful to
#' expand data frames to explicitly include missing data for countries.
#'
#' @inherit who_member_states return
#'
#' @export
oecd_member_states <- function() {
  x <- whoville::countries[["iso3"]]
  x[is_oecd_member(x)]
}

#' Check GBD high-income classification status from ISO3 codes.
#'
#' `is_gbd_high_income()` takes in a vector of ISO3 codes and returns a logical vector
#' on whether that country is classified as high-income in the 2019 Global Burden
#' of Disease released by the Institute for Health Metrics and Evaluation.
#'
#' @inheritParams is_who_member
#'
#' @return Logical vector.
#'
#' @export
is_gbd_high_income <- function(iso3) {
  members <- whoville::countries[["gbd_high_income"]]
  idx <- match(iso3, whoville::countries[["iso3"]])
  members[idx] %in% TRUE
}

#' Get ISO3 codes for GBD high-income states.
#'
#' `gbd_high_income_states()` returns ISO3 codes for states classified as high
#' income in the 2019 Global Burden of Disease released by the Institute for
#' Health Metrics and Evaluation. Useful to expand data frames to explicitly
#' include missing data for countries.
#'
#' @inherit who_member_states return
#'
#' @export
gbd_high_income_states <- function() {
  x <- whoville::countries[["iso3"]]
  x[is_gbd_high_income(x)]
}

#' Get ISO3 codes for least-developed countries.
#'
#' `un_ldcs()` returns ISO3 codes for countries classified as least
#' developed by the United Nations. Useful to expand data frames to explicitly
#' include missing data for countries.
#'
#' @inherit who_member_states return
#'
#' @export
un_ldcs <- function() {
  x <- whoville::countries[["iso3"]]
  x[is_un_ldc(x)]
}

#' Check least-developed country classification status from ISO3 codes.
#'
#' `is_un_ldc()` takes in a vector of ISO3 codes and returns a logical vector
#' on whether that country is classified as a least-developed country by the
#' United Nations.
#'
#' @inheritParams is_who_member
#'
#' @return Logical vector.
#'
#' @export
is_un_ldc <- function(iso3) {
  members <- whoville::countries[["un_ldc"]]
  idx <- match(iso3, whoville::countries[["iso3"]])
  members[idx] %in% TRUE
}

#' Get ISO3 codes for land-locked developing countries.
#'
#' `un_lldcs()` returns ISO3 codes for countries classified as land-locked
#' developing countries by the United Nations. Useful to expand data frames to explicitly
#' include missing data for countries.
#'
#' @inherit who_member_states return
#'
#' @export
un_lldcs <- function() {
  x <- whoville::countries[["iso3"]]
  x[is_un_lldc(x)]
}

#' Check land-locked developing country classification status from ISO3 codes.
#'
#' `is_un_lldc()` takes in a vector of ISO3 codes and returns a logical vector
#' on whether that country is classified as a land-locked developing country by the
#' United Nations.
#'
#' @inheritParams is_who_member
#'
#' @return Logical vector.
#'
#' @export
is_un_lldc <- function(iso3) {
  members <- whoville::countries[["un_lldc"]]
  idx <- match(iso3, whoville::countries[["iso3"]])
  members[idx] %in% TRUE
}

#' Get ISO3 codes for small island developing states.
#'
#' `un_sids()` returns ISO3 codes for countries classified as small island
#' developing states by the United Nations. Useful to expand data frames to explicitly
#' include missing data for countries.
#'
#' @inherit who_member_states return
#'
#' @export
un_sids <- function() {
  x <- whoville::countries[["iso3"]]
  x[is_un_sid(x)]
}

#' Check small island developing state status from ISO3 codes.
#'
#' `is_un_sid()` takes in a vector of ISO3 codes and returns a logical vector
#' on whether that country is classified as a small island developing state by the
#' United Nations.
#'
#' @inheritParams is_who_member
#'
#' @return Logical vector.
#'
#' @export
is_un_sid <- function(iso3) {
  members <- whoville::countries[["un_sids"]]
  idx <- match(iso3, whoville::countries[["iso3"]])
  members[idx] %in% TRUE
}

